pkg/kubernetes/scripts/virtual-machines/components/VmActions.jsx

```diff
@@ -0,0 +1,78 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// @flow
+
+import React, { PropTypes } from 'react';
+import { gettext as _ } from 'cockpit';
+import { connect } from 'react-redux'
+
+import type { Vm } from '../types.jsx';
+import { vmIdPrefx } from '../utils.jsx';
+import { vmShutdown } from '../kube-middleware.jsx';
+import { vmActionFailed } from '../action-creators.jsx';
+
+// import DropdownButtons from '../../../../machines/components/dropdownButtons.jsx';
+import { mouseClick } from '../utils.jsx';
+
+React;
+
+function canShutdown({ vm }: { vm: Vm }) {
+  vm ; // TODO: check on state
+  return true;
+}
+
+const VmActions = ({ vm, onFailure }: { vm: Vm, onFailure: Function }) => {
+  const id = vmIdPrefx(vm);
+
+  const onShutdown = () => {
+    vmShutdown({ vm }).catch((error) => {
+      console.info('VmActions: shutdown failed: ', error);
+      onFailure({
+        message: _("VM Shut Down failed."),
+        detail: error,
+      });
+    });
+  };
+
+  let shutdown = null;
+  if (canShutdown({ vm })) {
+    shutdown = (
+      <button className="btn btn-default btn-danger" onClick={mouseClick(onShutdown)} id={`${id}-shutdown`}>
+        {_("Shut Down")}
+      </button>);
+  }
+
+  return (
+    <div>
+      {shutdown}
+    </div>);
+}
+
+VmActions.propTypes = {
+    vm: PropTypes.object.isRequired,
+    onFailure: PropTypes.func.isRequired,
+};
+
+export default connect(
+  () => ({  }),
```

Comment:
What does that do?

Reply:
https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options

Simply said: the `connect` function injects redux store state&dispatch into a react component.
It's a way how to loose wiring of a react component on particular data provider (redux store) while relying purely on its props.
As a nice side-effect, a lot of ugly passing of state data and the `dispatch` function reference from higher to lower-level react components can be avoided.

In this particular case, there's no need to inject a portion of the state into this component, but the `connect` function expects first argument.